nameko/rpc.py

```diff
@@ -607,9 +630,21 @@ def _call(self, *args, **kwargs):
 
         return rpc_call
 
+    def _consolidate_timeouts(self, timeout, delivery_timeout):
+        if delivery_timeout is not None and timeout is None:
+            raise ValueError(
+                '`delivery_timeout` is given. '
+                'An explicit non-null `timeout` argument is required.'
+            )
+        if timeout is not None and delivery_timeout is None:
+            # If a timeout is given, we might as well remove the message from
+            # the server.
```

Comment:
Not sure this is valid. It is semantically different for the client to say _I no longer care about the result_ and _I don't care if this work is carried out_.

Reply:
Yes, I guess this is correct.  I'll take it out, thanks for catching it.  Along those lines:  is the naming ok (I have both of them named "timeout" to make clear their related, but we could also name it "expiration" is that is essentially what "delivery_timeout" does, pass that argument to the "expiration" header of rabbitmq.

Reply:
gRPC calls them client and server deadlines. I think in gRPC the server will actually abort a request in progress if the deadline is reached though.

I quite like "expiration" honestly. It lines up with the AMQP name for what happens to the message, and the server timeout will only be actioned here if the message is still queued in the broker when it's reached.

Reply:
Ok :+1:, unfortunately while working on this I stumbled on a few more tricky design decisions resulting in some more questions:

* "expiration" can already be set just by using the normal publisher options.  We could just leave that as it is.  The only disadvantages to this are:
  1. The option is for the entire rpc client (if you call different methods the same "expiration" would be set on all method calls).  
  2. The API might be a bit less intuitive especially once [Separate Publish & Consume options](https://www.notion.so/12116ff4a4b14d1ba0cf52e2faa84ea9?v=22ddef13d2774fedb53e9eea7572187a&p=83a2aefead1d45318b341913897f4cd0) get done.

In order to have both timeout and expiration to be set per call, we would need to have the options within the client.  Currently, my idea would be to have a seperate "publish_options" argument that would be passed to the `Client` object (for expiration), and pass the "timeout" argument to the client as well.  If one wanted to change the timeout per call, one would need to do something like:

```python
class TestService:
    name = 'test'
    other_service = ServiceRpc('other_service, timeout=2.0, publisher_options={'expiration': 5.0})
    
    def run(self):
        other_service.method1()  # called with expiration 5 and timeout 2
        method2 = other_service.method2
        method2.publish_kwargs['expiration'] = 10.0
        method2.timeout = 5
        method2()  # called with expiration 10.0, timeout 2
```

which is fine (even if it looks a bit ugly) ... but we can also cause a deadlock doing this since the "publisher_options" set in the main service are not passed the the client and we can set an expiration without setting a timeout, causing the call to wait forever:

```python
class TestService:
    name = 'test'
    other_service = ServiceRpc('other_service, timeout=2.0, publisher_options={'expiration': 5.0})
    
    def run(self):
        other_service.method1()  # called with expiration 5 and timeout 2
        method2 = other_service.method2
        method2.timeout = None
        method2()  # called with expiration 5.0 as set be the publisher, the client does not know this, so it will allow timeout to be set to `None`
```

The only way I can think of to work around this is to not allow the publisher options to be bound to the publisher at all and ALWAYS set them when "publish" is actually called, (which would allow us to check consistency when the call occurs).

Alternatively we can just not support setting timeouts per call and bind timeout and expiration to the `ClusterProxy` level and check everything there.

Reply:
Yes, this is a good point.

I think we should try to get the API right and then work from there.

I would want the existing publishers to continue to allow "expiration" to be passed. Both as a default when declaring it and optionally per-message.

The RPC client should have its own API and we should perhaps delineate the ones that apply to the underlying publisher. Something like:

```
client = RpcClient("service", timeout=10, expiration=20, publisher_options={}, consumer_options={})
```

`publisher_options` applies to the client publisher
`consumer_options` applies to the reply listener consumer

(It's arguable whether being able to change the consumer options of the reply listener is a good idea or not, but I'll leave it here for now)

We'd need to decide what to do when `publisher_options` includes `expiration` -- either ignore, warn or override the other value. I would probably warn and ignore.

Ideally every option could be provided at declaration time and also at method-call time. There is a general problem of how to do that with the RPC client though, because args and kwargs are both passed through to the remote function. In hindsight this was a bad design decision. As a workaround I generally favour using a context manager. Something like:

```
client = RpcClient("service", timeout=10, expiration=20, publisher_options={}, consumer_options={})

client.method()  # 10 second timeout
with RpcClient.override(client, timeout=20) as client:
    client.method()  # 20 second timeout
```

To avoid the deadlocked client issue I would just check at method-call time whether you've set an expiry without a client timeout, and raise an exception if so. If we were really clever we'd raise the exception at the end of the expiration period, if the call hadn't returned. That way the only effect of the safety mechanism would be that you'd get a a helpful exception in the client rather than waiting forever.

Reply:
Regarding 

```python
client = RpcClient("service", timeout=10, expiration=20, publisher_options={}, **publisher_options_kwargs)
```

I tried to implement this as suggested:
* `publisher_options_kwargs` still works but is now deprecated.  If we agree on everything I'd change the tests to use the new api accordingly as well.  
* `timeout` and `expiration` are explicit options and can be changed at any time up until a method call (they are basically passed all the way to the `rpc.Client` object).
* I did not implement the consumer options yet as it seems like a separate problem.


> We'd need to decide what to do when publisher_options includes expiration -- either ignore, warn or override the other value. I would probably warn and ignore.

Did this as suggested (warn and ignore).

> Ideally every option could be provided at declaration time and also at method-call time. [...] As a workaround I generally favor using a context manager.

Playing around with this I feel like a context manager is a bit clumsy since we're just changing attributes of the client object, but then creating a new client object anyway when we select a method.  Instead I thought a simple method might be simpler, so the current version implements the following api:

```python
self.service.method()  # normal method call
self.service.with_options(timeout=1).method()  # using timeout 1
```

the `with_options` name could also perhaps be changed to `using` resulting in something like `self.service.using(timeout=1).method()` ?

> To avoid the deadlocked client issue I would just check at method-call time whether you've set an expiry without a client timeout, and raise an exception if so

Did this

> If we were really clever we'd raise the exception at the end of the expiration period, if the call hadn't returned

Considering your previous comment about not implicitly setting the timeout if it's left out and thinking about this a bit I think it would be better to raise the exception directly at call time since I think it implies that the caller probably misconfigured the client and it's better to know this early (earliest point that this is the case is call time, since up until that time it could be changed).
Open to other interpretations (don't think it would be impossible to raise the exception after the expiration has run out). 

### Final tricky questions 

One thing that is still a bit tricky is this:

Obviously most of the `publisher_options` apply to the underlying publisher but can just as well apply to the `publish` method (where they then override the publisher options), so we are explicitly deciding that `expiration` can be set at call time but `priority`  and `headers` can not. 

If we would want to do something like:

```python
self.service.with_options(timeout=2, headers={'a': 3}, priority=3).method()
```

would change the api to:

```python
client = RpcClient("service", timeout=10, expiration=20, headers=None, priority=None, publisher_options={}, **publisher_options_kwargs)
```
??